### PR TITLE
On sided endpoint rework and onesided endpoint creation

### PR DESCRIPTION
**Pull request by MattBBaker**
The Endpoint rework/new flags that may impact others.
There is a function `create_one_sided_ep()` that can create endpoints from a UCXAddress (or a buffer with an UCXAddress) without the asyncio loop
The Endpoint class is modified to make the control/send/receive tags optional since the one sided EP won't have a pair (it only communicates through the worker)
The ApplicationContext at init asks for features UCP_FEATURE_RMA | UCP_FEATURE_AMO32 | UCP_FEATURE_AMO64 since these will be needed for `mem_map()` and one sided operations

### CODE REVIEW

**Comment by quasiben** [comment]
Do you mind if we target 0.17 for this ?  RAPIDS is in code freeze and I'd like to push anything to the  next branch

**Comment by MattBBaker** [comment]
I'm fine with this, how long will that be?

**Comment by quasiben** [comment]
Release is scheduler for Oct. 21st but we will nightly packages for the upcoming release as well as future work like this.  I'll make the change now

**Comment by pentschev** [comment]
I think @MattBBaker was asking when will 0.17 be released, and it's currently scheduled for December 9th.

**Comment by jakirkham** [comment]
Also we post the release schedule [here]( https://docs.rapids.ai/maintainers ).

**Comment by jakirkham** [comment]
For context, we are entering code freeze today for 0.16.0. Hence why we are asking about moving things to the next release to cutdown on churn while OPS finalizing things so we can ship out the stack. There's a fair bit of work involved in the release process and we want to avoid adding to it if we can.

**Comment by MattBBaker** [comment]
Okay, I understand. Anyways, I'm feeling good about this patch set now, who should we ping to make sure that the changes are not breaking anyone's code?

File: ucp/core.py
Comment (madsbk) [inline ucp/core.py:529]: I suggest that we unify all the tags attributes in a `dict` given as argument to `__init__()` and can be `None` when tags isn't used.

File: ucp/_libs/ucx_api.pyx
Comment (madsbk) [inline ucp/_libs/ucx_api.pyx:232]: I think we should make the options optional. `__init__()` could accept a set of future constants and default to:
```python
def __init__(self, config_dict, ucp_features={UCP_FEATURE_TAG, UCP_FEATURE_WAKEUP):
```
or 

```python
def __init__(self, config_dict, ucp_features={"UCP_FEATURE_TAG", "UCP_FEATURE_WAKEUP"):
```

**Review by madsbk — CHANGES_REQUESTED** [review]
It looks good @MattBBaker, I have some minor issues only.

File: ucp/_libs/ucx_api.pyx
Comment (MattBBaker) [reply ucp/_libs/ucx_api.pyx:232]: I was actually thinking about a mechanism like that. How would the features be passed in? If the strings are used then that would involve a lookup table/dict() while the named version would involve the user doing something like features=[ucp.feature.wakup, ucp.feature.tag]. Which one would be prefered?

File: ucp/core.py
Comment (MattBBaker) [reply ucp/core.py:529]: Any thoughts about adding tags after the fact? I do that on my test for convince, though I'm not sure I would use it in practice

File: ucp/_libs/ucx_api.pyx
Comment (madsbk) [reply ucp/_libs/ucx_api.pyx:232]: I have no strong opinion on this but I prefer it as string with a lookup dict.

File: ucp/core.py
Comment (madsbk) [reply ucp/core.py:529]: I don't think we should support _tags after the fact_ unless there is a real use case.

**Comment by MattBBaker** [comment]
@quasiben @madsbk @jakirkham Looks like the bots are all happy now, ready for the next round of review

File: ucp/core.py
Comment (madsbk) [inline ucp/core.py:538]: What about keeping the dict of tags like:
```suggestion
        self._tags = tags        
```
Than `self._tags==None` means _no tags_

File: ucp/core.py
Comment (madsbk) [inline ucp/core.py:576]:
```suggestion
            if self._tags:
```

**Review by madsbk — CHANGES_REQUESTED** [review]
I have some minor suggestions

File: ucp/core.py
Comment (madsbk) [reply ucp/core.py:576]: You will also have to change all the other `self._msg_tag_*` and `self._ctrl_tag_*` accesses  to dict lookup

File: ucp/core.py
Comment (MattBBaker) [reply ucp/core.py:401]: Yes, I should add the documentation. I wanted to be able to wire up an EP to still use the higher level tag API if you have a mechanism to distribute tags OOB.

File: ucp/core.py
Comment (MattBBaker) [reply ucp/core.py:538]: I did think about that, the main though is that it would require an additional dict lookup for every tag operation, which my MPI thinking brain says, "4 byte sends are valid, and some people care a lot about that kind of injection rate"

Comment (madsbk) [reply ucp/core.py:538]: I think we can afford dict lookup in this case. Alternatively, we could create a `struct` of the tag values that is `None` when tags aren't used.

**Comment by MattBBaker** [comment]
Whoops, tried updating my PR and something went wrong. Will work on

**Comment by jakirkham** [comment]
No worries. Is it ok if we retarget to `branch-0.18`? RAPIDS 0.17 is already shipped. Incidentally this may fix the same PR updating issue as well.

**Comment by jakirkham** [comment]
> No worries. Is it ok if we retarget to `branch-0.18`? RAPIDS 0.17 is already shipped. Incidentally this may fix the same PR updating issue as well.

Went ahead and made this change. Hope that is ok 🙂

Looks like this has cleaned up the commit history issue

**Comment by jakirkham** [comment]
Looks like `black` is running into an issue. Would you be able to run `black` locally and commit + push those changes? Should fix this issue

**Comment by madsbk** [comment]
> Looks like `black` is running into an issue. Would you be able to run `black` locally and commit + push those changes? Should fix this issue

Remember to use `black` version **19.10b0**

**Comment by MattBBaker** [comment]
Yea, it looks like I rebased on 0.18 by accident. And I swear I ran black before pushing this. I'll have a check on the version. and fix things up

**Comment by MattBBaker** [comment]
Turns out running black in the ucp directory will make it miss the test directory

**Comment by jakirkham** [comment]
JFYI I think all of us are out over the next couple weeks. Guessing you are also taking time off at some point. Thought you should know in case it gets a bit quiet here 🙂 Happy holidays btw! 🎉

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

**Comment by jakirkham** [comment]
Would you have another chance to look at this Mads? 🙂

File: ucp/core.py
Comment (madsbk) [inline ucp/core.py:238]:
```suggestion
```

**Review by madsbk — APPROVED** [review]
Great work @MattBBaker!
I only have a minor issue where you a duplicate `self.progress_tasks = []` line.
Thanks!

**Comment by MattBBaker** [comment]
Yea, I wanted to push something out because I figured it would be a little bit before I could circle back to this and wanted to push something out before the winter break

File: ucp/core.py
Comment (MattBBaker) [reply ucp/core.py:238]: Whoops. For some reason that has creeped in more than once.

File: ucp/_libs/ucx_api.pyx
Comment (jakirkham) [inline ucp/_libs/ucx_api.pyx:208]:
```suggestion
cdef dict ucp_features = {
```

**Review by jakirkham — APPROVED** [review]
Thanks Matt! LGTM as well. 😄 Made one minor suggestion below.

File: ucp/_libs/ucx_api.pyx
Comment (MattBBaker) [reply ucp/_libs/ucx_api.pyx]: This change seems good.

**Comment by MattBBaker** [comment]
I'm looking through the error messages, but I'm having a hard time interpreting what all they mean. I see a lot of time out errors but I see no indication of what is timing out and why.

**Comment by pentschev** [comment]
> I'm looking through the error messages, but I'm having a hard time interpreting what all they mean. I see a lot of time out errors but I see no indication of what is timing out and why.

The errors are legit, I checked locally with branch-0.18 and this PR and it's only reproduced with this PR. I did a little digging but I couldn't find what's the reason either. At first I thought these were a consequence of the new flags introduced in the `UCXContext` creation, but after reverting those, I still see the same issues. I'm really not sure what else could be the cause, after reviewing the PR again I don't see any other fundamental changes, but reverting both `ucp/core.py` and `ucp/endpoint_reuse.py` does resolve the issue, while keeping the new flags (but pretty much erasing what this PR introduces).

**Comment by MattBBaker** [comment]
I've been digging into the source of failures and I have come up with something interesting but I'm not sure how to proceed. Specifically in one of the new tag matching tests, `test_no_tag()` I get this trace back that is new
```sys:1: RuntimeWarning: coroutine 'BlockingMode._arm_worker' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Task was destroyed but it is pending!
task: <Task cancelling coro=<BlockingMode._arm_worker() running at /home/shadesfox/ucx-py/ucp/continuous_ucx_progress.py:88>>```
I'm not sure how/why this warning happens, but I suspect it is related to the timeouts, since it looks like the worker never gets armed. Can anyone help me figure out why this happens? 
@pentschev @jakirkham @madsbk

**Comment by pentschev** [comment]
I actually have seen this sort of issue over the past few weeks for certain tests/conditions, and I checked now that it's been happening since January 23rd with my nightly runs on `test_get_ucx_version`, plus it happens with some of dask-cuda benchmarks when there's something else running on the same machine causing an RMM OOM error, thus it indicates some sort of cleanup issue in that case. I remember I've seen it under other conditions (I can't rememeber exactly what, since I've been running dozens of different combinations lately) and had little success debugging it (ah, how much I love Python's async...).

What version of UCX are you running? UCX 1.9 is still ok with UCX-Py, but 1.10 and above are very flaky and we're working with UCX devs to fix some of those issues, so I would advise against anything above 1.9.

Could you also try running with https://github.com/rapidsai/ucx-py/pull/671 ? Generally speaking, that's a necessary fix, but I've seen hangs with it in some cases which I haven't managed to figure out why yet.

**Comment by MattBBaker** [comment]
I tried it with the patch, it seems to help but the problem doesn't completely go away

File: ucp/_libs/ucx_api.pyx
Comment (vchuravy) [inline ucp/_libs/ucx_api.pyx:246]: Why are the feature flags listed in the `field_mask`?

Comment (pentschev) [reply ucp/_libs/ucx_api.pyx:246]: Good catch @vchuravy, looks like a mistake to me. Could you check that @MattBBaker ?

**Comment by madsbk** [comment]
I think this PR is getting a bit too complex to merge easily. I will try to get all feature here implemented using a divide-and-conquer approach starting with https://github.com/rapidsai/ucx-py/pull/688.
I hope that is okay @MattBBaker?

**Comment by MattBBaker** [comment]
@madsbk I am 100% in favor of this. This was getting quite unwieldy anyways. And hopefully it will help narrow down what makes Dask especially unhappy

**Comment by madsbk** [comment]
@MattBBaker, with https://github.com/rapidsai/ucx-py/pull/691, we now support the creation of an endpoints from a `UCXAddress`. What is the next step? 
This PR brings this functionally to the _asyncio-level_ API through `core.create_endpoint_sync()` and `create_one_sided_ep()` but I am wondering if we should let it be a low-level non-asyncio feature for now?

Some context; the overall plan is to split UCX-Py into a high-level asyncio module and a low-level Cython module (`ucx_api.pyx`). In the next couple of weeks, I will work on making `ucx_api.pyx` self-contained with its own tests and docs.

**Comment by MattBBaker** [comment]
@madsbk What i the diff between what is in now and this PR?

Overall I'm in favor of allowing this to be a low level feature. For short term planning, the next step is to include the put/get APIs. On the subject of the split between the low level and high level modules, what is the plan to upstream work to the main UCX repo? That plan will inform the rest of the plan on how to fill in more functionality.

**Comment by madsbk** [comment]
> @madsbk What i the diff between what is in now and this PR?

It doesn't include the changes to `ucp/core.py` like `create_endpoint_sync` and `create_one_sided_ep`.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by MattBBaker** [comment]
Okay, so right now it live in the lower layer with no exposure to the higher level ucp? In that case it might be best to hold off for now on trying to get the higher level API in and let it remain a low level API.

**Comment by kkraus14** [comment]
add to allowlist

**Comment by madsbk** [comment]
> Okay, so right now it live in the lower layer with no exposure to the higher level ucp? In that case it might be best to hold off for now on trying to get the higher level API in and let it remain a low level API.

Yes exactly, let's implement some examples that make use of the one-sided low level API and get some experience before move to the high level API.

I am closing this PR but let's use the remaining features not already in `branch-0.19` as the bases for future work. Thanks for all the work @MattBBaker!

**Comment by jakirkham** [comment]
Would it make sense to target `branch-0.20` for this work?

**Comment by madsbk** [comment]
> Would it make sense to target `branch-0.20` for this work?

Yes the work should target future versions but I think it is better to submit new PRs that cherry-picks from this PR instead of trying to get this PR merged.

**Comment by MattBBaker** [comment]
Yea, this PR has gotten beastly. I'll pull in more work from my previous RMA branch too and focus on this strategy of filling out the low level API

**Pull request closed without merging**
